### ISSUE TITLE
Error in type signature of `zio.stream.ZSink#optional`

### ISSUE THREAD

**Issue by mlangc** [zio/zio#1981]
The problem is best illustrated by the following snippet:
```scala
val stream = Stream.fromIterable(List(1))
val sink = Sink.identity[Int].optional

// Should type to IO[Nothing, Option[Int]] instead:
val maybeInt: IO[Unit, Option[Int]] = stream.run(sink)
```

As pointed out by @iravid, the there is an error in the type signature of
```scala
def ? : ZSink[R, E, A, A, Option[B]]
```
where `E` should be replaced with `Nothing`.

**Comment by regiskuckaertz** [comment]
I don't think so, e.g. `ZIO.fromOption[A](v: => Option[A]): IO[Unit, A]`. That's the same here, as a sink that didn't receive anything yet cannot produce anything, so it must fail.

**Comment by iravid** [comment]
`ZSink#optional` is like `zio.fold(_ => None, Some(_))`, so it is supposed to eliminate the error type

**Comment by regiskuckaertz** [comment]
ok, I'm confused because that doesn't remove the `Unit` in `stream.run(sink)`. This is because `extractPure` cannot produce a `Right` for `ZSink.identity` if there never was an element in the first place.

**Comment by iravid** [comment]
Sorry can you show a complete sample?

**Comment by regiskuckaertz** [comment]
The one in the description in fact, the expression `stream.run(sink)` will still have type `IO[Unit, Option[Int]]`. Here's the console output:
```scala
scala> val stream = Stream.fromIterable(List(1))
stream: zio.stream.Stream[Nothing,Int] = zio.stream.StreamEffect@6ec407e4

scala> val sink = Sink.identity[Int].optional
sink: zio.stream.ZSink[Any,Nothing,Int,Int,Option[Int]] = zio.stream.ZSink$InputRemainderOps$$anon$15@1626d332

scala> val maybeInt: IO[Unit, Option[Int]] = stream.run(sink)
maybeInt: zio.IO[Unit,Option[Int]] = zio.ZIO$FlatMap@644ebdaa
```

I'm trying to find out why.

**Comment by regiskuckaertz** [comment]
```scala
scala> stream.transduce(sink)
res4: zio.stream.ZStream[Any,Nothing,Option[Int]] = zio.stream.ZStream@4abaa895

scala> stream.transduce(sink).runCollect
res5: zio.ZIO[Any,Nothing,List[Option[Int]]] = zio.ZIO$FlatMap@30e51de4
```
isn't the error type `Unit` because `run` can fail in the `Some(e)` case?

**Comment by regiskuckaertz** [comment]
yep:
```scala
scala> stream.process.use(identity)
res4: zio.ZIO[Any,Option[Nothing],Int] = zio.ZIO$CheckInterrupt@4b6f880d
```

**Comment by iravid** [comment]
Fixed in #1988
